### PR TITLE
MONGOID-5037 use $pull instead of $pullAll for when documents change before destroying all

### PR DESCRIPTION
**Pull request by Neilshweky**

### CODE REVIEW

**Comment by johnnyshields** [comment]
> This change relies on the assumption that all embedded documents have _id fields, which I believe they do.

@Neilshweky please see this ticket https://jira.mongodb.org/browse/MONGOID-5025 and my comments on it.

**Comment by johnnyshields** [comment]
What about doing something like `$pullAll => { "_id" => { "$exists" => true } }` ?

File: lib/mongoid/association/embedded/batchable.rb [hunk 1]
```diff
@@ -50,20 +50,39 @@ def batch_clear(docs)
...
         #
         # @example Batch remove the documents.
         #   batchable.batch_remove([ doc_one, doc_two ])
         #
         # @param [ Array<Document> ] docs The docs to remove.
         # @param [ Symbol ] method Delete or destroy.
         def batch_remove(docs, method = :delete)
-          removals = pre_process_batch_remove(docs, method)
+          # If the _id is nil, we cannot use $pull and delete by searching for
+          # the id. Therefore we have to user pullAll with the documents
```
Comment (p-mongo) [inline lib/mongoid/association/embedded/batchable.rb:62]:
```suggestion
          # the id. Therefore we have to user pullAll with the documents'
```

File: lib/mongoid/association/embedded/batchable.rb [hunk 2]
```diff
@@ -50,20 +50,39 @@ def batch_clear(docs)
...
+          # attributes.
+          pulls, pull_alls = pre_process_batch_remove(docs, method).partition do |o|
+            o.key?("_id") && !o["_id"].nil?
```
Comment (p-mongo) [inline lib/mongoid/association/embedded/batchable.rb:65]: The `key?` check here is redundant, no?

**Review by p-mongo — APPROVED** [review]

File: lib/mongoid/association/embedded/batchable.rb (hunk 1 above)
Comment (Neilshweky) [inline lib/mongoid/association/embedded/batchable.rb:62]:
```suggestion
          # the id. Therefore we have to use pullAll with the documents'
```

**Comment by johnnyshields** [comment]
@Neilshweky this PR looks good but just so we have a record here, what's the final outcome if a write of a non-_id doc arrives at the same time `.destroy_all` is called?

This case will of course be exceedingly rare, and not happen at all in standard usage of Mongoid. If users do expect it, they can use a **transaction** to mitigate the impact (i.e. the read which detects _id would be guaranteed to be synchronized with the $pullAll or $pull). Perhaps we can add a note in the documentation to that effect?

**Pull request merged**